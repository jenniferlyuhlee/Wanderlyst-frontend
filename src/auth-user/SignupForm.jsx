import React, { useState, useContext} from "react";
import { useNavigate, Navigate } from "react-router-dom";
import UserContext from "./UserContext";
import AlertCard from "../shared/Alert";

/** Signup Form component
 * Props: signup
 * - When submitted, calls signup function prop
 * - Redirects user to user homepage
 */

function SignupForm(){
    const initialState = {
        username: "",
        password: "",
        email: "",
        firstName: "",
        lastName: ""
    }
    // hooks
    const navigate = useNavigate();

    // states: formData, formErrors
    const [formData, setFormData] = useState(initialState);
    const [formErrors, setFormErrors] = useState([]);

    // general callback to update targeted form data
    function handleChange(e){
        const { name, value } = e.target;
        setFormData(data => ({...data, [name]: value}))
    }

    // form submission handler
    async function handleSubmit(e){
        e.preventDefault();
        alert(`signup successful: ${formData.username}`)
        // let result = await signup(formData);
        // if(result.sucess){
        //     navigate("/")
        // }
        // else{
        //     setFormErrors(result.err);
        // }
    }

  
    return(
        <div className = "Home">
            <div className = "container card px-5 auth-form"
                style = {{width: '32rem'}}>
                <h1 className = "text-center mb-4">Signup</h1>
                <p>Join now to share & uncover unforgettable experiences.</p>
                <form onSubmit = {handleSubmit} className = "row">
                    <div className = "form-floating mb-3 p-0">
                        <input id = "firstName"
                            className = "form-control"
                            name = "firstName"
                            placeholder = "firstName"
                            value = {formData.firstName}
                            onChange = {handleChange}
                        />
                        <label htmlFor="firstName">First Name</label>
                    </div>
                    <div className = "form-floating mb-3 p-0">
                        <input id = "lastName"
                            className = "form-control"
                            name = "lastName"
                            placeholder = "lastName"
                            value = {formData.lastName}
                            onChange = {handleChange}
                        />
                        <label htmlFor = "lastName">Last Name</label>
                    </div>
                    <div className = "form-floating mb-3 p-0">
                        <input
                            className = "form-control"
                            name = "email"
                            placeholder = "email"
                            value = {formData.email}
                            onChange = {handleChange}
                        />
                        <label htmlFor = "email">Email</label>
                    </div>
                    <div className = "form-floating mb-3 p-0">
                        <input
                            className = "form-control"
                            name = "username"
                            placeholder = "username"
                            value = {formData.username}
                            onChange = {handleChange}
                        />
                        <label htmlFor = "username">Username</label>
                    </div>
                    <div className = "form-floating mb-3 p-0">
                        <input
                            className = "form-control"
                            name = "password"
                            type = "password"
                            placeholder = "password"
                            value = {formData.password}
                            onChange = {handleChange}
                        />
                        <label htmlFor = "password">Password</label>
                    </div>
                    <button className = "btn rounded-pill btn-lg btn-primary">
                        Signup
                    </button>
                </form>
            </div>
        </div>
    )
}

export default SignupForm;