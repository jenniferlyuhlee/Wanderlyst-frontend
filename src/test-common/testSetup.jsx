import React, { useState } from "react";
import UserContext from "../auth-user/UserContext";
import WanderlystApi from "../utils/api";
import '@testing-library/jest-dom';

const demoUser = {
  username: "testadmin", 
  firstName: "testfirst",
  lastName: "testlast",
  location: "testloc",
  bio: "testbio",
  profilePic: null,
  isAdmin: true,
};

const demoTags = [
  {id: 1, name: 'Foodie', description: 'Savor the world one bite at a time. Dive into a journey packed with delicious local cuisines, hidden culinary gems, and must-try dishes that will leave your taste buds tingling.'},
  {id: 2, name: 'Explorer', description: 'Step off the beaten path and uncover the unknown. From ancient ruins to vibrant cityscapes, let your curiosity guide you through unforgettable experiences.'}
]

const UserProvider = ({ children, 
                        currUser = demoUser, 
                        setCurrUser, 
                        tags = demoTags,
                        logout,
                        hasLikedItin,
                        toggleLike }) => {
  const [user, setUser] = useState(currUser);

  return (
    <UserContext.Provider value={{ currUser: user, 
                                  setCurrUser: setUser, 
                                  tags, logout, 
                                  hasLikedItin, toggleLike}}>
      {children}
    </UserContext.Provider>
  );
};

export { UserProvider };