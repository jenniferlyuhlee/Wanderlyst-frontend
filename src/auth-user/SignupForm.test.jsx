import React from "react";
import { render, fireEvent, waitFor } from "@testing-library/react";
import { UserProvider } from "../test-common/testSetup";
import { MemoryRouter } from "react-router-dom";
import SignupForm from "./SignupForm";
import { beforeEach, expect, describe, test } from "vitest";

const mockSignup = vi.fn().mockResolvedValue({success: true});

describe("SignupForm Component", () => {
    beforeEach(()=> {
        mockSignup.mockClear();
    });

    test("renders without crashing: anon", () => {
        render(
        <MemoryRouter>
            <UserProvider currUser = {null}>
                <SignupForm signup={mockSignup} />
            </UserProvider>
        </MemoryRouter>)
    });
            
    
    test("matches snapshot", () => {
        const {asFragment} = render(
                            <MemoryRouter>
                                <UserProvider currUser = {null}>
                                    <SignupForm signup={mockSignup}/>
                                </UserProvider>
                            </MemoryRouter>)
        expect(asFragment()).toMatchSnapshot();
    });

    test("calls passed function with inputted values", async () => {
        const { getByRole, getByLabelText } = render(
            <MemoryRouter>
                <UserProvider currUser = {null}>
                    <SignupForm signup={mockSignup} />
                </UserProvider>
            </MemoryRouter>);
   
        // capture fields and login button
        const firstName = getByLabelText("First Name");
        const lastName = getByLabelText("Last Name");
        const email = getByLabelText("Email");
        const username = getByLabelText("Username");
        const password = getByLabelText("Password");
        const btn = getByRole("button");

        // fill in fields and submit
        fireEvent.change(firstName, {target: {value: 'testfirst'}});
        fireEvent.change(lastName, {target: {value: 'testlast'}});
        fireEvent.change(email, {target: {value: 'test@email.com'}});
        fireEvent.change(username, {target: {value: 'testuser'}});
        fireEvent.change(password, {target: {value: 'password'}});
        fireEvent.click(btn);
        
        await waitFor(() => expect(mockSignup).toHaveBeenCalled());
        await waitFor(() => expect(mockSignup).toHaveBeenCalledWith({
          firstName: 'testfirst',
          lastName: 'testlast',
          email: 'test@email.com',
          username: 'testuser',
          password: 'password'
        }));

    });
})