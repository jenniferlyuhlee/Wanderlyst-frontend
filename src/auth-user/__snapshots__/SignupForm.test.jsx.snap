// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`LoginForm Component > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="Home"
  >
    <div
      class="container card px-5 auth-form"
    >
      <h1
        class="text-center mb-3"
      >
        Signup
      </h1>
      <p
        class="my-2"
      >
        Join now to share & uncover unforgettable experiences.
      </p>
      <form
        class="row"
      >
        <div
          class="form-floating mb-3 p-0"
        >
          <input
            class="form-control"
            id="firstName"
            name="firstName"
            placeholder="firstName"
            required=""
            value=""
          />
          <label
            for="firstName"
          >
            First Name
          </label>
        </div>
        <div
          class="form-floating mb-3 p-0"
        >
          <input
            class="form-control"
            id="lastName"
            name="lastName"
            placeholder="lastName"
            required=""
            value=""
          />
          <label
            for="lastName"
          >
            Last Name
          </label>
        </div>
        <div
          class="form-floating mb-3 p-0"
        >
          <input
            class="form-control"
            id="email"
            name="email"
            placeholder="email"
            required=""
            value=""
          />
          <label
            for="email"
          >
            Email
          </label>
        </div>
        <div
          class="form-floating mb-3 p-0"
        >
          <input
            class="form-control"
            id="username"
            name="username"
            placeholder="username"
            required=""
            value=""
          />
          <label
            for="username"
          >
            Username
          </label>
        </div>
        <div
          class="form-floating mb-3 p-0"
        >
          <input
            class="form-control"
            id="password"
            name="password"
            placeholder="password"
            required=""
            type="password"
            value=""
          />
          <label
            for="password"
          >
            Password
          </label>
        </div>
        <button
          class="btn rounded-pill btn-lg btn-primary"
        >
          Signup
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Signup Form Component > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="Home"
  >
    <div
      class="container card px-5 auth-form"
    >
      <h1
        class="text-center mb-3"
      >
        Signup
      </h1>
      <p
        class="my-2"
      >
        Join now to share & uncover unforgettable experiences.
      </p>
      <form
        class="row"
      >
        <div
          class="form-floating mb-3 p-0"
        >
          <input
            class="form-control"
            id="firstName"
            name="firstName"
            placeholder="firstName"
            required=""
            value=""
          />
          <label
            for="firstName"
          >
            First Name
          </label>
        </div>
        <div
          class="form-floating mb-3 p-0"
        >
          <input
            class="form-control"
            id="lastName"
            name="lastName"
            placeholder="lastName"
            required=""
            value=""
          />
          <label
            for="lastName"
          >
            Last Name
          </label>
        </div>
        <div
          class="form-floating mb-3 p-0"
        >
          <input
            class="form-control"
            id="email"
            name="email"
            placeholder="email"
            required=""
            value=""
          />
          <label
            for="email"
          >
            Email
          </label>
        </div>
        <div
          class="form-floating mb-3 p-0"
        >
          <input
            class="form-control"
            id="username"
            name="username"
            placeholder="username"
            required=""
            value=""
          />
          <label
            for="username"
          >
            Username
          </label>
        </div>
        <div
          class="form-floating mb-3 p-0"
        >
          <input
            class="form-control"
            id="password"
            name="password"
            placeholder="password"
            required=""
            type="password"
            value=""
          />
          <label
            for="password"
          >
            Password
          </label>
        </div>
        <button
          class="btn rounded-pill btn-lg btn-primary"
        >
          Signup
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;

exports[`SignupForm Component > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="Home"
  >
    <div
      class="container card px-5 auth-form"
    >
      <h1
        class="text-center mb-3"
      >
        Signup
      </h1>
      <p
        class="my-2"
      >
        Join now to share & uncover unforgettable experiences.
      </p>
      <form
        class="row"
      >
        <div
          class="form-floating mb-3 p-0"
        >
          <input
            class="form-control"
            id="firstName"
            name="firstName"
            placeholder="firstName"
            required=""
            value=""
          />
          <label
            for="firstName"
          >
            First Name
          </label>
        </div>
        <div
          class="form-floating mb-3 p-0"
        >
          <input
            class="form-control"
            id="lastName"
            name="lastName"
            placeholder="lastName"
            required=""
            value=""
          />
          <label
            for="lastName"
          >
            Last Name
          </label>
        </div>
        <div
          class="form-floating mb-3 p-0"
        >
          <input
            class="form-control"
            id="email"
            name="email"
            placeholder="email"
            required=""
            value=""
          />
          <label
            for="email"
          >
            Email
          </label>
        </div>
        <div
          class="form-floating mb-3 p-0"
        >
          <input
            class="form-control"
            id="username"
            name="username"
            placeholder="username"
            required=""
            value=""
          />
          <label
            for="username"
          >
            Username
          </label>
        </div>
        <div
          class="form-floating mb-3 p-0"
        >
          <input
            class="form-control"
            id="password"
            name="password"
            placeholder="password"
            required=""
            type="password"
            value=""
          />
          <label
            for="password"
          >
            Password
          </label>
        </div>
        <button
          class="btn rounded-pill btn-lg btn-primary"
        >
          Signup
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
